src/outputs/alsa.c

```diff
@@ -1185,6 +1185,97 @@ alsa_write(struct output_buffer *obuf)
     }
 }
 
+// walk all the alsa devices here and report valid playback mixers
+static void
+alsa_device_list()
+{
+  snd_ctl_t *hdl = NULL;
+  snd_ctl_card_info_t *info = NULL;
+
+  snd_mixer_t *mixer = NULL;
+  snd_mixer_elem_t*  elem;
+  snd_mixer_selem_id_t* sid;
+
+  char hwdev[14];  // hw: (3) + max_uint (10)
+  int  err;
+  int  card = 0;
+  bool pbenabled = false;
+
+  snd_ctl_card_info_alloca(&info);
+  snd_mixer_selem_id_alloca(&sid);
+
+  while (card >= 0)
+    {
+      snprintf(hwdev, sizeof(hwdev), "hw:%d", card);
+
+      err = snd_ctl_open(&hdl, hwdev, 0);
+      if (err < 0)
+	{
+	  DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ASLA card=%d - %s\n", card, snd_strerror(err));
+	  if (snd_card_next(&card) < 0) {
+	    return;
+	  }
+	  continue;
+	}
+
+      err = snd_ctl_card_info(hdl, info);
+      if (err < 0)
+	{
+	  DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ASLA (info) card=%d - %s\n", card, snd_strerror(err));
+	  goto next;
```

Comment:
goto's should only be used for exit on error cleanup. When you need to use it like this it is a sign that the function is too large. So here you should make a function that dumps info about a single card and then call it from a loop.